File: skills/wechat-proxy-server/wechat_client.py
"""
微信 API 封装层：token 管理、图片上传、草稿创建
此模块运行在代理服务器上，拥有固定 IP，已加入微信白名单。
"""

import json
import os
import re
import time
from typing import Optional

import requests

_TOKEN_CACHE: dict = {}  # {appid: {token, expires_at}}
_TOKEN_CACHE_FILE = "/tmp/wechat_token_cache.json"


def _load_cache() -> dict:
    if os.path.exists(_TOKEN_CACHE_FILE):
        try:
            with open(_TOKEN_CACHE_FILE, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception:
            pass
    return {}


def _save_cache(cache: dict) -> None:
    try:
        with open(_TOKEN_CACHE_FILE, "w", encoding="utf-8") as f:
            json.dump(cache, f, ensure_ascii=False)
    except Exception:
        pass


def get_access_token(appid: str, appsecret: str) -> str:
    cache = _load_cache()
    entry = cache.get(appid, {})
    if entry.get("token") and entry.get("expires_at", 0) > time.time() + 300:
        return entry["token"]

    resp = requests.get(
        "https://api.weixin.qq.com/cgi-bin/token",
        params={"grant_type": "client_credential", "appid": appid, "secret": appsecret},
        timeout=15,
    )
    data = resp.json()
    if "access_token" not in data:
        raise Exception(f"获取 access_token 失败: {data.get('errmsg', str(data))}")

    token = data["access_token"]
    cache[appid] = {"token": token, "expires_at": time.time() + data.get("expires_in", 7200) - 300}
    _save_cache(cache)
    return token


_MIME_MAP = {
    "jpg": "image/jpeg", "jpeg": "image/jpeg",
    "png": "image/png", "gif": "image/gif", "webp": "image/webp",
}


def _guess_mime(url: str, content_type: str = "") -> tuple[str, str]:
    ext = url.rstrip("/").rsplit(".", 1)[-1].lower().split("?")[0]
    if ext in _MIME_MAP:
        return _MIME_MAP[ext], ext
    for e, m in _MIME_MAP.items():
        if content_type and m in content_type:
            return m, e
    return "image/png", "png"


def download_image(url: str, timeout: int = 30) -> tuple[bytes, str]:
    resp = requests.get(url, timeout=timeout, headers={"User-Agent": "Mozilla/5.0"})
    if resp.status_code != 200:
        raise Exception(f"下载失败: HTTP {resp.status_code}")
    return resp.content, resp.headers.get("Content-Type", "")


def upload_content_image(token: str, image_data: bytes, filename: str, mime: str) -> str:
    """上传正文图片（uploadimg 接口），返回 https URL。"""
    resp = requests.post(
        f"https://api.weixin.qq.com/cgi-bin/media/uploadimg?access_token={token}",
        files={"media": (filename, image_data, mime)},
        timeout=60,
    )
    result = resp.json()
    if "url" not in result:
        raise Exception(f"上传正文图片失败: {result}")
    url = result["url"]
    return url.replace("http://", "https://", 1) if url.startswith("http://") else url


def upload_permanent_image(token: str, image_data: bytes, filename: str = "cover.jpg") -> str:
    """上传永久素材图片，返回 media_id（用于封面）。"""
    resp = requests.post(
        f"https://api.weixin.qq.com/cgi-bin/material/add_material?access_token={token}&type=image",
        files={"media": (filename, image_data, "image/jpeg")},
        timeout=60,
    )
    result = resp.json()
    if "media_id" not in result:
        raise Exception(f"上传封面图失败: {result}")
    return result["media_id"]


def upload_permanent_video(
    token: str,
    video_data: bytes,
    filename: str = "video.mp4",
    title: str = "",
    introduction: str = "",
) -> dict:
    """
    上传永久视频素材 (add_material?type=video)。
    返回 { media_id, url }; url 可用于图文 content 中的 <video src=...>。
    - description 字段必须是 JSON 字符串: {"title": ..., "introduction": ...}
    - 仅认证服务号可用; 单个公众号最多 1000 条非图文永久素材
    - 视频建议 ≤ 10MB(部分文档允许 1GB,但实际中 10MB 最稳), 格式 mp4
    """
    if not video_data:
        raise Exception("视频文件为空")
    description = json.dumps(
        {"title": title or filename, "introduction": introduction or ""},
        ensure_ascii=False,
    )
    ext = (filename.rsplit(".", 1)[-1] or "mp4").lower().split("?")[0]
    mime = "video/mp4" if ext == "mp4" else ("video/quicktime" if ext == "mov" else "video/mp4")
    resp = requests.post(
        f"https://api.weixin.qq.com/cgi-bin/material/add_material?access_token={token}&type=video",
        files={"media": (filename, video_data, mime)},
        data={"description": description},
        timeout=300,  # 视频大, 给足超时
    )
    result = resp.json()
    if "media_id" not in result:
        raise Exception(f"上传视频素材失败: {result}")
    return {"media_id": result["media_id"], "url": result.get("url", "")}


def convert_video_to_mass(
    token: str,
    media_id: str,
    title: str = "",
    description: str = "",
) -> dict:
    """
    把永久素材 media_id 转成「可群发」的群发素材 media_id
    调 /cgi-bin/media/uploadvideo,POST JSON {media_id, title, description}
    title ≤ 64 字符, description ≤ 300 字符
    返回 { media_id, url }
    """
    if not media_id:
        raise Exception("media_id 必填")
    payload = {
        "media_id": media_id,
        "title": (title or "")[:64],
        "description": (description or "")[:300],
    }
    resp = requests.post(
        f"https://api.weixin.qq.com/cgi-bin/media/uploadvideo?access_token={token}",
        data=json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        headers={"Content-Type": "application/json; charset=utf-8"},
        timeout=30,
    )
    result = resp.json()
    if "media_id" not in result:
        raise Exception(f"转换群发素材失败: {result}")
    return {"media_id": result["media_id"], "url": result.get("url", "")}


def create_mass_video_task(
    token: str,
    mass_media_id: str,
    title: str = "",
    description: str = "",
    is_to_all: bool = True,
    tag_id: int = 0,
) -> dict:
    """
    创建 mpvideo 群发任务
    调 /cgi-bin/message/mass/sendall, msgtype=mpvideo
    群发后会进入 48 小时预览期,需在公众号后台手动点"群发"才会真正推送
    返回 { msg_id, msg_data_id }
    """
    if not mass_media_id:
        raise Exception("群发素材 media_id 必填")
    payload = {
        "filter": {"is_to_all": is_to_all} if is_to_all else {"is_to_all": False, "tag_id": tag_id},
        "mpvideo": {
            "media_id": mass_media_id,
            "title": (title or "")[:64],
            "description": (description or "")[:300],
        },
        "msgtype": "mpvideo",
        "send_ignore_reprint": 0,
    }
    resp = requests.post(
        f"https://api.weixin.qq.com/cgi-bin/message/mass/sendall?access_token={token}",
        data=json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        headers={"Content-Type": "application/json; charset=utf-8"},
        timeout=30,
    )
    result = resp.json()
    if result.get("errcode"):
        raise Exception(f"创建群发任务失败: {result}")
    return {
        "msg_id": result.get("msg_id", ""),
        "msg_data_id": result.get("msg_data_id", ""),
    }


def get_fallback_thumb_media_id(token: str) -> str:
    """从素材库取第一张图片的 media_id 作为封面 fallback。"""
    resp = requests.post(
        f"https://api.weixin.qq.com/cgi-bin/material/batchget_material?access_token={token}",
        json={"type": "image", "offset": 0, "count": 1},
        timeout=15,
    )
    items = resp.json().get("item", [])
    return items[0]["media_id"] if items else ""


def process_content_images(token: str, html: str) -> tuple[str, list[dict]]:
    """
    将 HTML 中所有外链图片下载后上传到微信，替换为 data-src 微信 URL。
    返回 (处理后的 HTML, 失败列表[{url, error}])
    """
    urls = list(dict.fromkeys(re.findall(r'<img[^>]+src="([^"]+)"[^>]*>', html)))
    if not urls:
        return html, []

    url_map: dict[str, str] = {}
    failures: list[dict] = []

    for i, url in enumerate(urls):
        try:
            data, ct = download_image(url)
            mime, ext = _guess_mime(url, ct)
            wechat_url = upload_content_image(token, data, f"img_{i}.{ext}", mime)
            url_map[url] = wechat_url
        except Exception as e:
            failures.append({"url": url, "error": str(e)})

    for orig, wx in url_map.items():
        html = re.sub(
            rf'<img([^>]*)src="{re.escape(orig)}"([^>]*)>',
            rf'<img\1data-src="{wx}"\2>',
            html,
        )

    return html, failures


def create_draft(
    token: str,
    title: str,
    content: str,
    thumb_media_id: str = "",
    content_source_url: str = "",
    digest: str = "",
) -> dict:
    """创建草稿，返回微信 API 响应。"""
    payload = {
        "articles": [{
            "title": title,
            "content": content,
            "thumb_media_id": thumb_media_id,
            "content_source_url": content_source_url,
            "digest": digest,
            "author": "",
            "need_open_comment": 1,
            "only_fans_can_comment": 0,
        }]
    }
    resp = requests.post(
        f"https://api.weixin.qq.com/cgi-bin/draft/add?access_token={token}",
        data=json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        headers={"Content-Type": "application/json; charset=utf-8"},
        timeout=30,
    )
    return resp.json()

File: skills/wechat-proxy-server/wechat_client.js
/**
 * 微信 API 封装层 (Node.js 版)
 * - token 管理（文件缓存）
 * - 图片下载与上传（正文图、永久素材封面）
 * - 草稿创建
 * - HTML 正文图片外链 -> 微信 CDN 替换
 *
 * 零依赖，纯 Node.js (>=18)
 */

const https = require('https');
const http = require('http');
const fs = require('fs');
const { URL } = require('url');

const TOKEN_CACHE_FILE = '/tmp/wechat_token_cache.json';

let _tokenCache = null;

function loadCache() {
  if (_tokenCache) return _tokenCache;
  try {
    if (fs.existsSync(TOKEN_CACHE_FILE)) {
      const raw = fs.readFileSync(TOKEN_CACHE_FILE, 'utf8');
      _tokenCache = JSON.parse(raw);
      return _tokenCache;
    }
  } catch (e) {
    // ignore corrupt cache
  }
  _tokenCache = {};
  return _tokenCache;
}

function saveCache(cache) {
  try {
    fs.writeFileSync(TOKEN_CACHE_FILE, JSON.stringify(cache, null, 2), 'utf8');
  } catch (e) {
    // ignore fs errors (e.g. /tmp readonly in some containers)
  }
  _tokenCache = cache;
}

const MIME_MAP = {
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  png: 'image/png',
  gif: 'image/gif',
  webp: 'image/webp',
};

function guessMime(url, contentType = '') {
  const clean = url.split('?')[0].split('#')[0];
  const ext = clean.split('.').pop().toLowerCase();
  if (MIME_MAP[ext]) {
    return { mime: MIME_MAP[ext], ext };
  }
  for (const [e, m] of Object.entries(MIME_MAP)) {
    if (contentType && contentType.toLowerCase().includes(e)) {
      return { mime: m, ext: e };
    }
  }
  return { mime: 'image/png', ext: 'png' };
}

// 通用 HTTPS 请求封装
async function httpsRequest(method, urlStr, { body = null, headers = {} } = {}) {
  return new Promise((resolve, reject) => {
    const u = new URL(urlStr);
    const client = u.protocol === 'https:' ? https : http;

    const opts = {
      hostname: u.hostname,
      port: u.port || (u.protocol === 'https:' ? 443 : 80),
      path: u.pathname + u.search,
      method: method.toUpperCase(),
      headers: { ...headers },
    };

    if (body) {
      if (Buffer.isBuffer(body)) {
        opts.headers['Content-Length'] = body.length;
      } else if (typeof body === 'string' || body instanceof Uint8Array) {
        opts.headers['Content-Length'] = Buffer.byteLength(body);
      }
    }

    const req = client.request(opts, (res) => {
      const chunks = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end', () => {
        const buf = Buffer.concat(chunks);
        const text = buf.toString('utf8');
        const ct = (res.headers['content-type'] || '').toLowerCase();

        let data;
        // 优先按 Content-Type 解析;微信 add_material?type=video 接口偶尔返回
        // Content-Type: text/plain 但 body 实际是 JSON,做兜底
        if (ct.includes('application/json')) {
          try {
            data = JSON.parse(text);
          } catch {
            data = { raw: text };
          }
        } else if (ct.includes('text/plain') || ct === '') {
          try {
            data = JSON.parse(text);
          } catch {
            data = { raw: text, contentType: ct };
          }
        } else {
          data = { raw: text, contentType: ct };
        }

        if (res.statusCode >= 400) {
          const err = new Error(`HTTP ${res.statusCode}: ${text.slice(0, 300)}`);
          err.status = res.statusCode;
          err.response = data;
          reject(err);
        } else {
          resolve(data);
        }
      });
    });

    req.on('error', reject);

    if (body) {
      req.write(body);
    }
    req.end();
  });
}

async function httpsGetJSON(urlStr) {
  return httpsRequest('GET', urlStr);
}

async function httpsPostJSON(urlStr, payload) {
  const jsonStr = JSON.stringify(payload);
  return httpsRequest('POST', urlStr, {
    body: jsonStr,
    headers: { 'Content-Type': 'application/json; charset=utf-8' },
  });
}

function buildMultipart(fieldName, filename, mimeType, buffer) {
  const boundary = '----NodeFormBoundary' + Date.now().toString(16) + Math.random().toString(16).slice(2);
  const header = Buffer.from(
    `--${boundary}\r\n` +
      `Content-Disposition: form-data; name="${fieldName}"; filename="${filename}"\r\n` +
      `Content-Type: ${mimeType}\r\n\r\n`
  );
  const footer = Buffer.from(`\r\n--${boundary}--\r\n`);
  const body = Buffer.concat([header, buffer, footer]);
  return { boundary, body };
}

// 构造多字段 multipart/form-data (用于上传到微信 add_material?type=video
// — 需要同时携带 media(文件) 和 description(JSON 字符串))
function buildMultipartFields(fields, fileBoundary) {
  const boundary = fileBoundary || ('----NodeFormBoundary' + Date.now().toString(16) + Math.random().toString(16).slice(2));
  const parts = [];
  for (const f of fields) {
    let header;
    if (f.filename) {
      header =
        `--${boundary}\r\n` +
        `Content-Disposition: form-data; name="${f.name}"; filename="${f.filename}"\r\n` +
        `Content-Type: ${f.contentType || 'application/octet-stream'}\r\n\r\n`;
    } else {
      header =
        `--${boundary}\r\n` +
        `Content-Disposition: form-data; name="${f.name}"\r\n\r\n`;
    }
    parts.push(Buffer.from(header));
    parts.push(Buffer.isBuffer(f.data) ? f.data : Buffer.from(String(f.data), 'utf8'));
    parts.push(Buffer.from('\r\n'));
  }
  parts.push(Buffer.from(`--${boundary}--\r\n`));
  return { boundary, body: Buffer.concat(parts) };
}

// 极简 multipart/form-data 解析器(用于本代理服务器接收客户端上传的视频文件)
// 仅支持 RFC 7578 常见形式(单文件 + 若干普通字段),不处理嵌套/多文件
function parseMultipart(buffer, contentType) {
  const m = /boundary=(?:"([^"]+)"|([^;]+))/i.exec(contentType || '');
  if (!m) throw new Error('multipart 请求缺少 boundary');
  const boundary = '--' + (m[1] || m[2]).trim();
  const text = buffer.toString('binary');
  const parts = [];
  const segments = text.split(boundary);
  // segments: ['', header1\r\n\r\nbody1\r\n, header2\r\n\r\nbody2\r\n, --]
  for (let i = 1; i < segments.length; i++) {
    const seg = segments[i];
    if (seg === '--' || seg === '--\r\n') break;
    const idx = seg.indexOf('\r\n\r\n');
    if (idx === -1) continue;
    const head = seg.slice(0, idx);
    let body = seg.slice(idx + 4);
    if (body.endsWith('\r\n')) body = body.slice(0, -2);

    const nameM = /name="([^"]+)"/i.exec(head);
    const fileM = /filename="([^"]*)"/i.exec(head);
    const ctM = /Content-Type:\s*([^\r\n]+)/i.exec(head);
    if (!nameM) continue;

    const name = nameM[1];
    if (fileM) {
      parts.push({
        name,
        filename: fileM[1],
        contentType: ctM ? ctM[1].trim() : 'application/octet-stream',
        data: Buffer.from(body, 'binary'),
      });
    } else {
      parts.push({ name, data: Buffer.from(body, 'binary').toString('utf8') });
    }
  }
  return parts;
}

async function httpsPostMultipart(urlStr, boundary, buffer) {
  return httpsRequest('POST', urlStr, {
    body: buffer,
    headers: {
      'Content-Type': `multipart/form-data; boundary=${boundary}`,
    },
  });
}

// ==================== 公开 API ====================

async function getAccessToken(appid, appsecret) {
  const cache = loadCache();
  const entry = cache[appid] || {};
  if (entry.token && entry.expires_at > Math.floor(Date.now() / 1000) + 300) {
    return entry.token;
  }

  const url = `https://api.weixin.qq.com/cgi-bin/token?grant_type=client_credential&appid=${appid}&secret=${appsecret}`;
  const data = await httpsGetJSON(url);

  if (!data.access_token) {
    throw new Error(`获取 access_token 失败: ${data.errmsg || JSON.stringify(data)}`);
  }

  const token = data.access_token;
  const expires_in = data.expires_in || 7200;

  cache[appid] = {
    token,
    expires_at: Math.floor(Date.now() / 1000) + expires_in - 300,
  };
  saveCache(cache);
  return token;
}

async function downloadImage(url, timeoutMs = 30000) {
  return new Promise((resolve, reject) => {
    const u = new URL(url);
    const client = u.protocol === 'https:' ? https : http;

    const req = client.get(
      url,
      {
        headers: { 'User-Agent': 'Mozilla/5.0 (compatible; WeChatProxy/1.0)' },
        timeout: timeoutMs,
      },
      (res) => {
        if (res.statusCode !== 200) {
          reject(new Error(`下载失败: HTTP ${res.statusCode}`));
          res.resume();
          return;
        }
        const chunks = [];
        res.on('data', (c) => chunks.push(c));
        res.on('end', () => {
          resolve({
            buffer: Buffer.concat(chunks),
            contentType: res.headers['content-type'] || '',
          });
        });
      }
    );

    req.on('error', reject);
    req.on('timeout', () => {
      req.destroy();
      reject(new Error('下载超时'));
    });
  });
}

async function uploadContentImage(token, imageData, filename, mime) {
  const { boundary, body } = buildMultipart('media', filename, mime, imageData);
  const url = `https://api.weixin.qq.com/cgi-bin/media/uploadimg?access_token=${token}`;
  const result = await httpsPostMultipart(url, boundary, body);
  if (!result.url) {
    throw new Error(`上传正文图片失败: ${JSON.stringify(result)}`);
  }
  let u = result.url;
  if (typeof u === 'string' && u.startsWith('http://')) {
    u = 'https://' + u.slice(7);
  }
  return u;
}

async function uploadPermanentImage(token, imageData, filename = 'cover.jpg') {
  const ext = (filename.split('.').pop() || 'jpg').toLowerCase().split('?')[0];
  const mime = MIME_MAP[ext] || 'image/jpeg';
  const { boundary, body } = buildMultipart('media', filename, mime, imageData);
  const url = `https://api.weixin.qq.com/cgi-bin/material/add_material?access_token=${token}&type=image`;
  const result = await httpsPostMultipart(url, boundary, body);
  if (!result.media_id) {
    throw new Error(`上传封面图失败: ${JSON.stringify(result)}`);
  }
  return result.media_id;
}

// 上传永久视频素材 (material/add_material?type=video)
// 返回 { media_id, url } —— url 是微信返回的视频源地址,可用于图文 content 中 <video src=...>
// 注意:
//   - description 字段是 JSON 字符串: {"title": "...", "introduction": "..."}
//   - 仅认证服务号可用; 单个公众号最多 1000 条非图文永久素材
async function uploadPermanentVideo(token, videoBuffer, filename = 'video.mp4', title = '', introduction = '') {
  if (!Buffer.isBuffer(videoBuffer) || videoBuffer.length === 0) {
    throw new Error('视频文件为空');
  }
  const ext = (filename.split('.').pop() || 'mp4').toLowerCase().split('?')[0];
  const mime = ext === 'mp4' ? 'video/mp4' : (ext === 'mov' ? 'video/quicktime' : 'video/mp4');
  const description = JSON.stringify({
    title: title || filename,
    introduction: introduction || '',
  });
  const { boundary, body } = buildMultipartFields([
    { name: 'media', filename, contentType: mime, data: videoBuffer },
    { name: 'description', data: description },
  ]);
  const url = `https://api.weixin.qq.com/cgi-bin/material/add_material?access_token=${token}&type=video`;
  const result = await httpsPostMultipart(url, boundary, body);
  if (!result.media_id) {
    throw new Error(`上传视频素材失败: ${JSON.stringify(result)}`);
  }
  return {
    media_id: result.media_id,
    url: result.url || '',
  };
}

async function getFallbackThumbMediaId(token) {
  const url = `https://api.weixin.qq.com/cgi-bin/material/batchget_material?access_token=${token}`;
  const res = await httpsPostJSON(url, { type: 'image', offset: 0, count: 1 });
  const items = (res && res.item) || [];
  return items.length ? items[0].media_id : '';
}

async function processContentImages(token, html) {
  // 提取所有 img src（去重，保持顺序）
  const urls = [];
  const seen = new Set();
  const imgRegex = /<img[^>]+src=["']([^"']+)["'][^>]*>/gi;
  let match;
  while ((match = imgRegex.exec(html)) !== null) {
    const u = match[1];
    if (!seen.has(u)) {
      seen.add(u);
      urls.push(u);
    }
  }

  if (urls.length === 0) {
    return { html, failures: [] };
  }

  const urlMap = {};
  const failures = [];

  for (let i = 0; i < urls.length; i++) {
    const origUrl = urls[i];
    try {
      const { buffer, contentType } = await downloadImage(origUrl);
      const { mime, ext } = guessMime(origUrl, contentType);
      const wxUrl = await uploadContentImage(token, buffer, `img_${i}.${ext}`, mime);
      urlMap[origUrl] = wxUrl;
    } catch (e) {
      failures.push({ url: origUrl, error: e.message || String(e) });
    }
  }

  // 替换 src 为 data-src（仅替换成功上传的）
  let processed = html;
  for (const [orig, wx] of Object.entries(urlMap)) {
    const safe = orig.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const re = new RegExp(`(<img[^>]*?)src=["']${safe}["']([^>]*>)`, 'gi');
    processed = processed.replace(re, `$1data-src="${wx}"$2`);
  }

  return { html: processed, failures };
}

async function createDraft({
  token,
  title,
  content,
  thumb_media_id = '',
  content_source_url = '',
  digest = '',
}) {
  const payload = {
    articles: [
      {
        title,
        content,
        thumb_media_id,
        content_source_url,
        digest,
        author: '',
        need_open_comment: 1,
        only_fans_can_comment: 0,
      },
    ],
  };
  const url = `https://api.weixin.qq.com/cgi-bin/draft/add?access_token=${token}`;
  return httpsPostJSON(url, payload);
}

module.exports = {
  getAccessToken,
  downloadImage,
  uploadContentImage,
  uploadPermanentImage,
  uploadPermanentVideo,
  getFallbackThumbMediaId,
  processContentImages,
  createDraft,
  parseMultipart,
  // 内部工具按需导出（测试用）
  _guessMime: guessMime,
  _buildMultipartFields: buildMultipartFields,
};
